src/asr/asr_utils.py

```diff
@@ -235,6 +204,87 @@ def _plot_and_save_attention(self, att_w, filename):
         plt.close()
 
 
+def restore_snapshot(model, snapshot, load_fn=chainer.serializers.load_npz):
+    '''Extension to restore snapshot'''
+    @training.make_extension(trigger=(1, 'epoch'))
+    def restore_snapshot(trainer):
+        _restore_snapshot(model, snapshot, load_fn)
+
+    return restore_snapshot
+
+
+def _restore_snapshot(model, snapshot, load_fn=chainer.serializers.load_npz):
+    load_fn(snapshot, model)
+    logging.info('restored from ' + str(snapshot))
+
+
+def adadelta_eps_decay(eps_decay):
+    '''Extension to perform adadelta eps decay'''
+    @training.make_extension(trigger=(1, 'epoch'))
+    def adadelta_eps_decay(trainer):
+        _adadelta_eps_decay(trainer, eps_decay)
+
+    return adadelta_eps_decay
+
+
+def _adadelta_eps_decay(trainer, eps_decay):
+    optimizer = trainer.updater.get_optimizer('main')
+    # for chainer
+    if hasattr(optimizer, 'eps'):
+        current_eps = optimizer.eps
+        setattr(optimizer, 'eps', current_eps * eps_decay)
+        logging.info('adadelta eps decayed to ' + str(optimizer.eps))
+    # pytorch
+    else:
+        for p in optimizer.param_groups:
+            p["eps"] *= eps_decay
+            logging.info('adadelta eps decayed to ' + str(p["eps"]))
+
+
+def torch_snapshot(savefun=torch.save,
+                   filename='snapshot.ep.{.updater.epoch}'):
+    """Returns a trainer extension to take snapshots of the trainer for pytorch."""
+    @extension.make_extension(trigger=(1, 'epoch'), priority=-100)
+    def torch_snapshot(trainer):
+        _torch_snapshot_object(trainer, trainer, filename.format(trainer), savefun)
+
+    return torch_snapshot
+
+
+def _torch_snapshot_object(trainer, target, filename, savefun):
+    # make snapshot_dict dictionary
+    s = DictionarySerializer()
+    s.save(trainer)
+    if hasattr(trainer.updater.model, "model"):
+        # (for TTS)
+        if hasattr(trainer.updater.model.model, "module"):
```

Comment:
Why TTS has a bit redundant model structure (`model.model`)?
Is there a way to make it same as ASR?

Reply:
If we assume that same options are used when resuming, it will be more simple.

But if you want to keep same format as model.loss.best, current format is needed.